ddtrace/ext/ci_visibility/api.py

```diff
@@ -0,0 +1,435 @@
+"""
+Provides the API necessary to interacting with the CI Visibility service.
+
+NOTE: BETA - this API is currently in development and is subject to change.
+
+This API supports the agentless, and agent-proxied (EVP) modes. It does not support the APM protocol.
+
+All functions in this module are meant to be called in a stateless manner. Test runners (or custom implementations) that
+rely on this API are not expected to keep CI Visibility-related state for each session, module, suite or test.
+
+Stable values of module, suite, test names, and parameters, are a necessity for this API to function properly.
+
+All types and methods for interacting with the API are provided and documented in this file.
+"""
+import dataclasses
+from enum import Enum
+from types import TracebackType
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import NamedTuple
+from typing import Optional
+from typing import Tuple
+from typing import Type
+
+from ddtrace.ext.ci_visibility._ci_visibility_base import _CIVisibilityAPIBase
+from ddtrace.ext.ci_visibility._ci_visibility_base import _CIVisibilityChildItemIdBase
+from ddtrace.ext.ci_visibility._ci_visibility_base import _CIVisibilityRootItemIdBase
+from ddtrace.ext.ci_visibility.constants import DEFAULT_OPERATION_NAMES
+from ddtrace.ext.ci_visibility.constants import DEFAULT_SESSION_NAME
+from ddtrace.ext.ci_visibility.util import _catch_and_log_exceptions
+from ddtrace.ext.test import Status as TestStatus
+from ddtrace.internal import core
+from ddtrace.internal.core.event_hub import ResultType
+from ddtrace.internal.logger import get_logger
+
+
+log = get_logger(__name__)
+
+
+class CITestStatus(Enum):
+    PASS = TestStatus.PASS.value
+    FAIL = TestStatus.FAIL.value
+    SKIP = TestStatus.SKIP.value
+    XFAIL = TestStatus.XFAIL.value
+    XPASS = TestStatus.XPASS.value
```

Comment:
This is definitely goofy and will need cleaning up at some point (the `internal` pieces will likely import this instead of the other way around).